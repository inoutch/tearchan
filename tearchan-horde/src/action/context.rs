use crate::action::manager::TimeMilliseconds;

pub struct ActionContext {
    pub last_time: TimeMilliseconds, // The last time of stacking all actions
    pub running_end_time: TimeMilliseconds, // The end time of running action
    pub state_len: usize,            // Running and pending state length of entity
}
