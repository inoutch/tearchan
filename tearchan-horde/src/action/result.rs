use crate::action::Action;
use std::sync::Arc;

pub enum ActionResult<T> {
    Start {
        action: Arc<Action<T>>,
    },
    Update {
        action: Arc<Action<T>>,
        current_time: u64,
    },
    End {
        action: Arc<Action<T>>,
    },
}

impl<T> ActionResult<T> {
    pub fn get_start(&self) -> Option<&Arc<Action<T>>> {
        match self {
            ActionResult::Start { action } => Some(action),
            ActionResult::Update { .. } => None,
            ActionResult::End { .. } => None,
        }
    }

    pub fn get_update(&self) -> Option<(&Arc<Action<T>>, u64)> {
        match self {
            ActionResult::Start { .. } => None,
            ActionResult::Update {
                action,
                current_time,
            } => Some((action, *current_time)),
            ActionResult::End { .. } => None,
        }
    }

    pub fn get_end(&self) -> Option<&Arc<Action<T>>> {
        match self {
            ActionResult::Start { .. } => None,
            ActionResult::Update { .. } => None,
            ActionResult::End { action } => Some(action),
        }
    }
}
