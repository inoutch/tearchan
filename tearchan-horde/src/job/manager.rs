use crate::action::manager::{
    ActionClientManager, ActionController, ActionManager, ActionManagerData, ActionManagerTrait,
    ActionServerManager, TimeMilliseconds,
};
use crate::action::result::ActionResult;
use crate::HordeInterface;
use std::collections::VecDeque;
use std::ops::Deref;
use std::option::Option::Some;
use std::sync::Arc;
use tearchan_ecs::component::EntityId;

const MAX_LOOP_SIZE_PER_FRAME: usize = 1000usize;

pub struct JobManager<T>
where
    T: HordeInterface,
{
    action_manager: ActionManager<T::ActionState>,
}

impl<T> Default for JobManager<T>
where
    T: HordeInterface,
{
    fn default() -> Self {
        JobManager {
            action_manager: ActionManager::default(),
        }
    }
}

impl<T> JobManager<T>
where
    T: HordeInterface,
{
    pub fn run(&mut self, provider: &mut T, elapsed_time: TimeMilliseconds) {
        match &mut self.action_manager {
            ActionManager::Server(action_manager) => {
                action_manager.update(elapsed_time);

                let mut loop_count = 0;
                let mut next_action_time = action_manager.next_action_time();
                loop {
                    if next_action_time != action_manager.next_action_time() {
                        update_change_time(provider, action_manager);
                        next_action_time = action_manager.next_action_time();
                    }

                    let mut is_changed = update_action(provider, action_manager);
                    let entity_ids = action_manager.clean_pending_entity_ids();
                    for entity_id in entity_ids {
                        let mut priority = 0;
                        let mut job_queue: VecDeque<T::Job> = VecDeque::new();
                        job_queue.push_front(get_until_some_priority_is_returned(
                            |entity_id, priority| {
                                provider.on_first(entity_id, priority, &action_manager.reader())
                            },
                            entity_id,
                            &mut priority,
                        ));

                        while let Some(job) = job_queue.pop_front() {
                            let result = provider.on_next(entity_id, job, &action_manager.reader());
                            if job_queue.is_empty()
                                && result.states.is_empty()
                                && result.creators.is_empty()
                            {
                                // If the jobs and actions cannot be generated from the current job tree,
                                // change the priority and recreate the first job
                                priority += 1;
                                job_queue.push_front(get_until_some_priority_is_returned(
                                    |entity_id, priority| {
                                        provider.on_first(
                                            entity_id,
                                            priority,
                                            &action_manager.reader(),
                                        )
                                    },
                                    entity_id,
                                    &mut priority,
                                ));
                                continue;
                            }

                            // Update actions
                            is_changed |= !result.states.is_empty();

                            // Add action creators
                            for creator in result.creators.into_iter().rev() {
                                job_queue.push_front(creator);
                            }
                            // Add actions
                            action_manager.push_states(entity_id, result.states);
                        }
                    }

                    if !is_changed {
                        break;
                    }
                    loop_count += 1;
                    debug_assert!(loop_count < MAX_LOOP_SIZE_PER_FRAME);
                }
            }
            ActionManager::Client(action_manager) => {
                action_manager.update(elapsed_time);
                update_action_in_client(provider, action_manager);
            }
        }
    }

    pub fn action_controller(&mut self) -> ActionController<T::ActionState> {
        self.action_manager.controller()
    }

    pub fn action_manager(&self) -> &ActionManager<T::ActionState> {
        &self.action_manager
    }

    pub fn action_manager_mut(&mut self) -> &mut ActionManager<T::ActionState> {
        &mut self.action_manager
    }

    pub fn replace_action_manager(&mut self, data: ActionManagerData<T::ActionState>) {
        self.action_manager = match &self.action_manager {
            ActionManager::Server(_) => ActionManager::Server(ActionServerManager::new(data)),
            ActionManager::Client(_) => ActionManager::Client(ActionClientManager::new(data)),
        };
    }
}

impl<T: HordeInterface> From<ActionManager<T::ActionState>> for JobManager<T> {
    fn from(action_manager: ActionManager<T::ActionState>) -> Self {
        JobManager { action_manager }
    }
}

fn update_action<T>(
    provider: &mut T,
    action_manager: &mut ActionServerManager<T::ActionState>,
) -> bool
where
    T: HordeInterface,
{
    let next_action_time = action_manager.next_action_time();
    while next_action_time == action_manager.next_action_time() {
        let result = match action_manager.pull() {
            None => return false,
            Some(result) => result,
        };
        let mut controller = action_manager.controller();
        update_event(provider, &mut controller, result);
    }
    true
}

fn update_change_time<T>(provider: &mut T, action_manager: &mut ActionServerManager<T::ActionState>)
where
    T: HordeInterface,
{
    for (entity_id, state) in provider
        .on_change_time(&action_manager.reader())
        .into_iter()
    {
        action_manager.push_states(entity_id, vec![state]);
    }
}

fn update_event<T>(
    provider: &mut T,
    controller: &mut ActionController<T::ActionState>,
    result: ActionResult<T::ActionState>,
) where
    T: HordeInterface,
{
    match result {
        ActionResult::Start { action } => {
            provider.on_send(Arc::clone(&action));
            provider.on_start(action.deref(), controller);
        }
        ActionResult::Update {
            action,
            current_time,
        } => {
            let duration = action.end_time() - action.start_time();
            let ratio = (current_time - action.start_time()) as f32 / duration as f32;
            provider.on_update(action.deref(), ratio, controller);
        }
        ActionResult::End { action } => {
            provider.on_end(action.deref(), controller);
        }
        ActionResult::Cancel { action } => {
            provider.on_cancel(&action, controller);
        }
        ActionResult::Enqueue { action } => {
            provider.on_enqueue(&action, controller);
        }
    }
}

fn update_action_in_client<T, U>(provider: &mut T, action_manager: &mut U)
where
    T: HordeInterface,
    U: ActionManagerTrait<T::ActionState>,
{
    while let Some(result) = action_manager.pull() {
        let mut controller = action_manager.controller();
        match result {
            ActionResult::Start { action } => {
                provider.on_send(Arc::clone(&action));
                provider.on_start(action.deref(), &mut controller);
            }
            ActionResult::Update {
                action,
                current_time,
            } => {
                let duration = action.end_time() - action.start_time();
                let ratio = (current_time - action.start_time()) as f32 / duration as f32;
                provider.on_update(action.deref(), ratio, &mut controller);
            }
            ActionResult::End { action } => {
                provider.on_end(action.deref(), &mut controller);
            }
            ActionResult::Cancel { action } => {
                provider.on_cancel(&action, &mut controller);
            }
            ActionResult::Enqueue { action } => {
                provider.on_enqueue(&action, &mut controller);
            }
        }
    }
}

fn get_until_some_priority_is_returned<F, R>(mut f: F, entity_id: EntityId, priority: &mut u32) -> R
where
    F: FnMut(EntityId, u32) -> Option<R>,
{
    loop {
        if let Some(v) = f(entity_id, *priority) {
            return v;
        }
        *priority += 1;

        debug_assert!(*priority < 1000, "Priority has been exceeded");
    }
}

#[cfg(test)]
mod test {
    use crate::action::manager::{ActionController, ActionServerReader, TimeMilliseconds};
    use crate::action::Action;
    use crate::job::manager::JobManager;
    use crate::job::result::JobResult;
    use crate::{HordeInterface, ProgressState};
    use std::cell::RefCell;
    use tearchan_ecs::component::group::ComponentGroup;
    use tearchan_ecs::component::EntityId;
    use tearchan_util::id_manager::IdManager;

    #[derive(Debug)]
    enum Kind {
        Dog,
        Cat,
        Human,
    }

    #[derive(Debug, Clone)]
    #[allow(dead_code)]
    enum CustomActionState {
        Move { position: Position },
        Sleep,
        Eat { food_name: &'static str },
        Job { salary: u32 },
    }

    #[derive(Debug, Clone)]
    struct Position((u32, u32));

    #[derive(Debug, Clone)]
    #[allow(dead_code)]
    enum CustomActionCreator {
        EatLunch {
            position: Position,
            food_name: &'static str,
        },
        EatFood {
            food_name: &'static str,
        },
        MoveTo {
            position: Position,
        },
        Sleep,
        Job {
            salary: u32,
        },
        Work {
            salary: u32,
            position: Position,
        },
        Invalid,
    }

    #[derive(Debug)]
    #[allow(dead_code)]
    enum ActionKind {
        Start {
            current_time: TimeMilliseconds,
            action: Action<CustomActionState>,
        },
        End {
            current_time: TimeMilliseconds,
            action: Action<CustomActionState>,
        },
        Update {
            current_time: TimeMilliseconds,
            action: Action<CustomActionState>,
            ratio: f32,
        },
        Enqueue {
            action: Action<CustomActionState>,
        },
        First {
            entity_id: EntityId,
            priority: u32,
            current_time: TimeMilliseconds,
        },
        Next {
            entity_id: EntityId,
            job: CustomActionCreator,
            current_time: TimeMilliseconds,
        },
        Cancel {
            action: Action<CustomActionState>,
        },
        ChangeTime {
            current_time: TimeMilliseconds,
        },
    }

    struct CustomGame {
        pub entity_id_manager: IdManager<EntityId>,
        pub kind_components: ComponentGroup<Kind>,
        pub name_components: ComponentGroup<String>,
        pub position_components: ComponentGroup<Position>,
        pub actions: RefCell<Vec<ActionKind>>,
    }

    impl HordeInterface for CustomGame {
        type ActionState = CustomActionState;
        type Job = CustomActionCreator;

        fn on_start(
            &mut self,
            action: &Action<Self::ActionState>,
            _controller: &mut ActionController<CustomActionState>,
        ) {
            self.actions.borrow_mut().push(ActionKind::Start {
                current_time: action.start_time(),
                action: action.clone(),
            });
        }

        fn on_update(
            &mut self,
            action: &Action<Self::ActionState>,
            ratio: f32,
            _controller: &mut ActionController<CustomActionState>,
        ) {
            self.actions.borrow_mut().push(ActionKind::Update {
                current_time: (((action.end_time() - action.start_time()) as f32 * ratio)
                    as TimeMilliseconds
                    + action.start_time()),
                action: action.clone(),
                ratio,
            });
        }

        fn on_end(
            &mut self,
            action: &Action<Self::ActionState>,
            _controller: &mut ActionController<CustomActionState>,
        ) {
            self.actions.borrow_mut().push(ActionKind::End {
                current_time: action.end_time(),
                action: action.clone(),
            });
        }

        fn on_enqueue(
            &mut self,
            action: &Action<Self::ActionState>,
            _controller: &mut ActionController<CustomActionState>,
        ) {
            self.actions.borrow_mut().push(ActionKind::Enqueue {
                action: action.clone(),
            });
        }

        fn on_first(
            &self,
            entity_id: EntityId,
            priority: u32,
            reader: &ActionServerReader<Self::ActionState>,
        ) -> Option<Self::Job> {
            self.actions.borrow_mut().push(ActionKind::First {
                entity_id,
                priority,
                current_time: reader.current_time(),
            });
            let kind = self.kind_components.get(entity_id).unwrap();
            Some(match kind {
                Kind::Dog => CustomActionCreator::EatLunch {
                    position: Position((100, 200)),
                    food_name: "dog food",
                },
                Kind::Cat => CustomActionCreator::Sleep,
                Kind::Human => CustomActionCreator::Work {
                    position: Position((100, 200)),
                    salary: 200, // $200
                },
            })
        }

        fn on_next(
            &self,
            entity_id: EntityId,
            job: Self::Job,
            reader: &ActionServerReader<Self::ActionState>,
        ) -> JobResult<Self::Job, Self::ActionState> {
            self.actions.borrow_mut().push(ActionKind::Next {
                entity_id,
                job: job.clone(),
                current_time: reader.current_time(),
            });
            let mut result = JobResult::default();
            match job {
                CustomActionCreator::EatLunch {
                    position,
                    food_name,
                } => {
                    result
                        .creators
                        .push(CustomActionCreator::MoveTo { position });
                    result
                        .creators
                        .push(CustomActionCreator::EatFood { food_name });
                    return result;
                }
                CustomActionCreator::MoveTo { position } => {
                    result
                        .states
                        .push((CustomActionState::Move { position }, 1000));
                    return result;
                }
                CustomActionCreator::Sleep { .. } => {
                    result.states.push((CustomActionState::Sleep, 3000));
                    return result;
                }
                CustomActionCreator::Work { position, salary } => {
                    result
                        .creators
                        .push(CustomActionCreator::MoveTo { position });
                    result.creators.push(CustomActionCreator::Job { salary });
                    return result;
                }
                CustomActionCreator::EatFood { food_name } => {
                    result
                        .states
                        .push((CustomActionState::Eat { food_name }, 2000));
                    return result;
                }
                CustomActionCreator::Job { salary } => {
                    result
                        .states
                        .push((CustomActionState::Job { salary }, 5000));
                    return result;
                }
                _ => {}
            }

            JobResult {
                states: vec![(CustomActionState::Sleep, 1000)],
                creators: vec![],
            }
        }

        fn on_cancel(
            &mut self,
            action: &Action<Self::ActionState>,
            _controller: &mut ActionController<Self::ActionState>,
        ) {
            self.actions.borrow_mut().push(ActionKind::Cancel {
                action: action.clone(),
            });
        }

        fn on_change_time(
            &mut self,
            reader: &ActionServerReader<Self::ActionState>,
        ) -> Vec<(EntityId, ProgressState<Self::ActionState>)> {
            self.actions.borrow_mut().push(ActionKind::ChangeTime {
                current_time: reader.current_time(),
            });
            Vec::with_capacity(0)
        }
    }

    #[test]
    fn test_custom_game() {
        let entity_id_manager: IdManager<EntityId> = IdManager::new(0, |id| id + 1);
        let kind_components: ComponentGroup<Kind> = ComponentGroup::default();
        let name_components: ComponentGroup<String> = ComponentGroup::default();
        let position_components: ComponentGroup<Position> = ComponentGroup::default();

        let mut job_manager: JobManager<CustomGame> = JobManager::default();
        let mut custom_game = CustomGame {
            entity_id_manager,
            kind_components,
            name_components,
            position_components,
            actions: RefCell::new(Vec::new()),
        };

        // Create cat
        {
            let entity_id = custom_game.entity_id_manager.create_generator().gen();
            custom_game.kind_components.push(entity_id, Kind::Cat);
            custom_game
                .name_components
                .push(entity_id, "Melon".to_string());
            custom_game
                .position_components
                .push(entity_id, Position((0, 0)));
            job_manager.action_controller().attach(entity_id);
        }
        // Create dog
        {
            let entity_id = custom_game.entity_id_manager.create_generator().gen();
            custom_game.kind_components.push(entity_id, Kind::Dog);
            custom_game
                .name_components
                .push(entity_id, "Puffball".to_string());
            custom_game
                .position_components
                .push(entity_id, Position((30, 10)));
            job_manager.action_controller().attach(entity_id);
        }
        // Create human
        {
            let entity_id = custom_game.entity_id_manager.create_generator().gen();
            custom_game.kind_components.push(entity_id, Kind::Human);
            custom_game
                .name_components
                .push(entity_id, "Jack".to_string());
            custom_game
                .position_components
                .push(entity_id, Position((4, 8)));
            job_manager.action_controller().attach(entity_id);
        }

        // Process horde
        job_manager.run(&mut custom_game, 10000);

        insta::assert_debug_snapshot!("actions", custom_game.actions.borrow());
    }
}
