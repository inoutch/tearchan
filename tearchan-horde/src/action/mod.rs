use crate::action::manager::TimeMilliseconds;
use serde::{Deserialize, Serialize};
use std::rc::Rc;
use tearchan_ecs::component::EntityId;

pub mod context;
pub mod manager;
pub mod result;

#[derive(Debug, Serialize, Deserialize)]
pub struct Action<T> {
    entity_id: EntityId,
    start_time: TimeMilliseconds,
    end_time: TimeMilliseconds,
    inner: Rc<T>,
}

impl<T> Action<T> {
    pub fn new(
        entity_id: EntityId,
        start: TimeMilliseconds,
        end: TimeMilliseconds,
        inner: T,
    ) -> Action<T> {
        Action {
            entity_id,
            start_time: start,
            end_time: end,
            inner: Rc::new(inner),
        }
    }
}

impl<T> Action<T> {
    pub fn entity_id(&self) -> EntityId {
        self.entity_id
    }

    pub fn start_time(&self) -> TimeMilliseconds {
        self.start_time
    }

    pub fn end_time(&self) -> TimeMilliseconds {
        self.end_time
    }

    pub fn inner(&self) -> &Rc<T> {
        &self.inner
    }
}
