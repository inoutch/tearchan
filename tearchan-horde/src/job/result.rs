use crate::action::manager::TimeMilliseconds;

pub type ProgressState<ActionState> = (ActionState, TimeMilliseconds);

#[derive(Debug)]
pub struct JobResult<Job, ActionState> {
    pub creators: Vec<Job>,
    pub states: Vec<ProgressState<ActionState>>,
}

impl<ActionCreator, ActionState> Default for JobResult<ActionCreator, ActionState> {
    fn default() -> Self {
        JobResult {
            creators: vec![],
            states: vec![],
        }
    }
}
