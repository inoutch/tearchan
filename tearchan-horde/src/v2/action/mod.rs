use crate::action::manager::TimeMilliseconds;
use crate::v2::Tick;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tearchan_ecs::component::EntityId;

pub mod collection;
pub mod manager;

pub const VALID_SESSION_ID: ActionSessionId = 0;

pub type ActionSessionId = u64;

#[derive(Copy, Clone, Debug, Serialize, Deserialize)]
pub enum ActionType {
    Start {
        tick: Tick,
    },
    Update {
        start: TimeMilliseconds,
        end: TimeMilliseconds,
    },
    End {
        tick: Tick,
    },
}

impl ActionType {
    pub fn tick(&self) -> Option<Tick> {
        match self {
            ActionType::Start { tick, .. } => Some(*tick),
            ActionType::Update { .. } => None,
            ActionType::End { tick, .. } => Some(*tick),
        }
    }
}

#[derive(Debug, Serialize, Deserialize)]
pub struct Action<T> {
    raw: T,
    entity_id: EntityId,
    ty: ActionType,
}

impl<T> Action<T> {
    pub fn new(raw: T, entity_id: EntityId, ty: ActionType) -> Self {
        Self { raw, entity_id, ty }
    }

    pub fn tick(&self) -> Option<Tick> {
        self.ty.tick()
    }
}

pub type ArcAction<T> = Action<Arc<T>>;

impl<T> Action<T> {
    pub fn raw(&self) -> &T {
        &self.raw
    }

    pub fn replace<U>(&self, raw: U) -> Action<U> {
        Action {
            raw,
            entity_id: self.entity_id,
            ty: self.ty,
        }
    }

    pub fn entity_id(&self) -> EntityId {
        self.entity_id
    }
}
