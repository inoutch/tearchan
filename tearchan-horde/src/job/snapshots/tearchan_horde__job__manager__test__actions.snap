---
source: tearchan-horde/src/job/manager.rs
expression: custom_game.actions.borrow()

---
[
    First {
        entity_id: 0,
        priority: 0,
        current_time: 0,
    },
    Next {
        entity_id: 0,
        job: Sleep,
        current_time: 0,
    },
    First {
        entity_id: 1,
        priority: 0,
        current_time: 0,
    },
    Next {
        entity_id: 1,
        job: EatLunch {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
            food_name: "dog food",
        },
        current_time: 0,
    },
    Next {
        entity_id: 1,
        job: MoveTo {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 0,
    },
    Next {
        entity_id: 1,
        job: EatFood {
            food_name: "dog food",
        },
        current_time: 0,
    },
    First {
        entity_id: 2,
        priority: 0,
        current_time: 0,
    },
    Next {
        entity_id: 2,
        job: Work {
            salary: 200,
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 0,
    },
    Next {
        entity_id: 2,
        job: MoveTo {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 0,
    },
    Next {
        entity_id: 2,
        job: Job {
            salary: 200,
        },
        current_time: 0,
    },
    Enqueue {
        action: Action {
            entity_id: 0,
            start_time: 0,
            end_time: 3000,
            inner: Sleep,
        },
    },
    Start {
        current_time: 0,
        action: Action {
            entity_id: 0,
            start_time: 0,
            end_time: 3000,
            inner: Sleep,
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 1000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 0,
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 1000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 1000,
            end_time: 3000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 2,
            start_time: 0,
            end_time: 1000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 0,
        action: Action {
            entity_id: 2,
            start_time: 0,
            end_time: 1000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 2,
            start_time: 1000,
            end_time: 6000,
            inner: Job {
                salary: 200,
            },
        },
    },
    ChangeTime {
        current_time: 0,
    },
    End {
        current_time: 1000,
        action: Action {
            entity_id: 1,
            start_time: 0,
            end_time: 1000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 1000,
        action: Action {
            entity_id: 1,
            start_time: 1000,
            end_time: 3000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    End {
        current_time: 1000,
        action: Action {
            entity_id: 2,
            start_time: 0,
            end_time: 1000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 1000,
        action: Action {
            entity_id: 2,
            start_time: 1000,
            end_time: 6000,
            inner: Job {
                salary: 200,
            },
        },
    },
    ChangeTime {
        current_time: 1000,
    },
    End {
        current_time: 3000,
        action: Action {
            entity_id: 0,
            start_time: 0,
            end_time: 3000,
            inner: Sleep,
        },
    },
    End {
        current_time: 3000,
        action: Action {
            entity_id: 1,
            start_time: 1000,
            end_time: 3000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    First {
        entity_id: 0,
        priority: 0,
        current_time: 3000,
    },
    Next {
        entity_id: 0,
        job: Sleep,
        current_time: 3000,
    },
    First {
        entity_id: 1,
        priority: 0,
        current_time: 3000,
    },
    Next {
        entity_id: 1,
        job: EatLunch {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
            food_name: "dog food",
        },
        current_time: 3000,
    },
    Next {
        entity_id: 1,
        job: MoveTo {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 3000,
    },
    Next {
        entity_id: 1,
        job: EatFood {
            food_name: "dog food",
        },
        current_time: 3000,
    },
    Enqueue {
        action: Action {
            entity_id: 0,
            start_time: 3000,
            end_time: 6000,
            inner: Sleep,
        },
    },
    Start {
        current_time: 3000,
        action: Action {
            entity_id: 0,
            start_time: 3000,
            end_time: 6000,
            inner: Sleep,
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 3000,
            end_time: 4000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 3000,
        action: Action {
            entity_id: 1,
            start_time: 3000,
            end_time: 4000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 4000,
            end_time: 6000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    ChangeTime {
        current_time: 3000,
    },
    End {
        current_time: 4000,
        action: Action {
            entity_id: 1,
            start_time: 3000,
            end_time: 4000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 4000,
        action: Action {
            entity_id: 1,
            start_time: 4000,
            end_time: 6000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    ChangeTime {
        current_time: 4000,
    },
    End {
        current_time: 6000,
        action: Action {
            entity_id: 2,
            start_time: 1000,
            end_time: 6000,
            inner: Job {
                salary: 200,
            },
        },
    },
    End {
        current_time: 6000,
        action: Action {
            entity_id: 0,
            start_time: 3000,
            end_time: 6000,
            inner: Sleep,
        },
    },
    End {
        current_time: 6000,
        action: Action {
            entity_id: 1,
            start_time: 4000,
            end_time: 6000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    First {
        entity_id: 0,
        priority: 0,
        current_time: 6000,
    },
    Next {
        entity_id: 0,
        job: Sleep,
        current_time: 6000,
    },
    First {
        entity_id: 1,
        priority: 0,
        current_time: 6000,
    },
    Next {
        entity_id: 1,
        job: EatLunch {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
            food_name: "dog food",
        },
        current_time: 6000,
    },
    Next {
        entity_id: 1,
        job: MoveTo {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 6000,
    },
    Next {
        entity_id: 1,
        job: EatFood {
            food_name: "dog food",
        },
        current_time: 6000,
    },
    First {
        entity_id: 2,
        priority: 0,
        current_time: 6000,
    },
    Next {
        entity_id: 2,
        job: Work {
            salary: 200,
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 6000,
    },
    Next {
        entity_id: 2,
        job: MoveTo {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 6000,
    },
    Next {
        entity_id: 2,
        job: Job {
            salary: 200,
        },
        current_time: 6000,
    },
    Enqueue {
        action: Action {
            entity_id: 0,
            start_time: 6000,
            end_time: 9000,
            inner: Sleep,
        },
    },
    Start {
        current_time: 6000,
        action: Action {
            entity_id: 0,
            start_time: 6000,
            end_time: 9000,
            inner: Sleep,
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 6000,
            end_time: 7000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 6000,
        action: Action {
            entity_id: 1,
            start_time: 6000,
            end_time: 7000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 7000,
            end_time: 9000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 2,
            start_time: 6000,
            end_time: 7000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 6000,
        action: Action {
            entity_id: 2,
            start_time: 6000,
            end_time: 7000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 2,
            start_time: 7000,
            end_time: 12000,
            inner: Job {
                salary: 200,
            },
        },
    },
    ChangeTime {
        current_time: 6000,
    },
    End {
        current_time: 7000,
        action: Action {
            entity_id: 1,
            start_time: 6000,
            end_time: 7000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 7000,
        action: Action {
            entity_id: 1,
            start_time: 7000,
            end_time: 9000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    End {
        current_time: 7000,
        action: Action {
            entity_id: 2,
            start_time: 6000,
            end_time: 7000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 7000,
        action: Action {
            entity_id: 2,
            start_time: 7000,
            end_time: 12000,
            inner: Job {
                salary: 200,
            },
        },
    },
    ChangeTime {
        current_time: 7000,
    },
    End {
        current_time: 9000,
        action: Action {
            entity_id: 0,
            start_time: 6000,
            end_time: 9000,
            inner: Sleep,
        },
    },
    End {
        current_time: 9000,
        action: Action {
            entity_id: 1,
            start_time: 7000,
            end_time: 9000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    First {
        entity_id: 0,
        priority: 0,
        current_time: 9000,
    },
    Next {
        entity_id: 0,
        job: Sleep,
        current_time: 9000,
    },
    First {
        entity_id: 1,
        priority: 0,
        current_time: 9000,
    },
    Next {
        entity_id: 1,
        job: EatLunch {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
            food_name: "dog food",
        },
        current_time: 9000,
    },
    Next {
        entity_id: 1,
        job: MoveTo {
            position: Position(
                (
                    100,
                    200,
                ),
            ),
        },
        current_time: 9000,
    },
    Next {
        entity_id: 1,
        job: EatFood {
            food_name: "dog food",
        },
        current_time: 9000,
    },
    Enqueue {
        action: Action {
            entity_id: 0,
            start_time: 9000,
            end_time: 12000,
            inner: Sleep,
        },
    },
    Start {
        current_time: 9000,
        action: Action {
            entity_id: 0,
            start_time: 9000,
            end_time: 12000,
            inner: Sleep,
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 9000,
            end_time: 10000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 9000,
        action: Action {
            entity_id: 1,
            start_time: 9000,
            end_time: 10000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Enqueue {
        action: Action {
            entity_id: 1,
            start_time: 10000,
            end_time: 12000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    ChangeTime {
        current_time: 9000,
    },
    Update {
        current_time: 10000,
        action: Action {
            entity_id: 2,
            start_time: 7000,
            end_time: 12000,
            inner: Job {
                salary: 200,
            },
        },
        ratio: 0.6,
    },
    End {
        current_time: 10000,
        action: Action {
            entity_id: 1,
            start_time: 9000,
            end_time: 10000,
            inner: Move {
                position: Position(
                    (
                        100,
                        200,
                    ),
                ),
            },
        },
    },
    Start {
        current_time: 10000,
        action: Action {
            entity_id: 1,
            start_time: 10000,
            end_time: 12000,
            inner: Eat {
                food_name: "dog food",
            },
        },
    },
    Update {
        current_time: 10000,
        action: Action {
            entity_id: 0,
            start_time: 9000,
            end_time: 12000,
            inner: Sleep,
        },
        ratio: 0.33333334,
    },
    ChangeTime {
        current_time: 10000,
    },
]
